src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/srm/algebras/BV.cs

```diff
@@ -8,19 +8,26 @@
 namespace System.Text.RegularExpressions.SRM
 {
     /// <summary>
-    /// Represents a bitvector of given Length (number of bits).
+    /// Represents a bitvector of arbitrary length (i.e. number of bits).
```

Comment:
Could we use `BitArray` instead?  If not, what would it take to make `BitArray` suitable?

Reply:
Yes, I think `BitArray` should be used instead. The whole `BV` class could probably get removed and BVAlgebra could implement `ICharAlgebra<BitArray>` instead. @veanes do you remember if there's any code that depends on the exact representation of `BV`?

Reply:
No, BV is used instead of BV64 (i.e. ulong) when there are more than 64 parts in the partition (that almost never happens but could and there are unit tests that trigger its use). 
But other than that, the details of what the type actually is should never leak through. Replacing BV by BitArray implies reimplementing all the Boolean operations in the algebra: Bitwise-OR, Bitwise-AND, complement, etc... all the interface methods of the algebra. Then providing a convenienf pretty printer etc. ... So some amount of programming needed, but should be pretty straightforward.

**One important aspect being that BitArray can be use as a threadsafe valuetype (that might be a dealbraker here). I highly doubt also that is will be more efficient than BV. Usually BV will have say to (or at most 3) ulongs tat it consists ofe --- very low overhead**

Reply:
I looked at the datatype. I don't think it would be useful to replace BV. What concerns me most is that BitArray is not a valuetype. Adding wrappers that make it a valuetype would defeat the purpose. For example implementing bitwise conjunction. With BV it is essentially bitwise-AND of say two ulongs (say when you have < 128 parts, which is covers all the cases I can think of).

Reply:
I think the motivation is mostly to reduce the amount of code. The term type has to be used immutably, so memory allocation for operations is inevitable and I'd guess that's the most expensive thing here.

Reply:
I looked into the `BitArray` implementation and it's pretty much all the same logic as `BV`, except it uses 32-bit ints instead of 64-bit ones. It also does have code paths for vectorization, but those don't get triggered with less than 256 bits.

BitArray not being immutable is a concern. It would probably be best to have the actual term type be an opaque wrapper around BitArray that doesn't expose the mutating functions to other code, just so no-one in the future introduces code outside BVAlgebra that mutates the terms.

Reply:
OK. Then I leave the decision up to you. If adding the wrapper for the immutable type and the operations is worth the effort, I agree and **am all for** reducing code size (i.e. getting rid of BV). 
First, there would need to be operations such as C = A & B where A, B, C are BitArrays and C is the resulting bitwise-AND of A and B where neither A nor B is changed. Currently such operations are missing in the datatype.

Reply:
... making **BV be the immutable wrapper** and basing its implementation on BitArray, is what I would suggest then.

Reply:
... Perhaps renaming it BitVector

Reply:
> I think the motivation is mostly to reduce the amount of code

Yup.  My goal is to reduce duplication, and wherever possible, centralize on shared implementations rather than having one-offs that need to be maintained and that don't benefit from any improvements to that centralized thing.  Additionally, as use cases present themselves that the current implementation may not be suitable for, it can drive us to determine what would enable it to be suitable, and then augment the centralized thing, making it better for everyone.

Reply:
Just a note: The following two tests are (currently) the only two related unit test to be checked more closely later, that will be affected by this change: `SRMTest_BV `and `SRMTest_BV_WideLatin`. More unit tests could of course be added for more corner cases of BV. (e.g. adding a case reaching up to or more than 256 parts)